fitbenchmarking/core/results_output.py

```diff
@@ -199,6 +201,287 @@ def preprocess_data(results: "list[FittingResult]"):
     return best_results, sorted_results
 
 
+def _handle_fallback_tags(results,
+                          all_result_tags,
+                          fallback_columns,
+                          col_order):
+    """
+    The function that relabels the fallback column tags that
+    appear when there is jacobian and hessian fallbacks for a
+    minimizer.
+
+    :param results: The list of results of benchmarking
+    :type results: list[FittingResult]
+    :param all_result_tags: A list of tags that can be used
+                            to sort the results
+    :type all_result_tags: list[dict[str, str]]
+    :param fallback_columns: The col tag of the fallback columns
+    :type fallback_columns: list[str]
+    :param col_order: The sort order of the col tags
+    :type col_order: list[str]
+
+    :return: all results and the results tags
+    :rtype: list[FittingResult], list[dict[str, str]]
+    """
+    column_rename = "best_avaliable"
+
+    # Find column tags and count for each tag
+    column_summary = dict((k,
+                          [tag['col'] for tag in
+                           all_result_tags].count(k)) for
+                          k in {tag['col'] for
+                                tag in all_result_tags})
+
+    # Find software and minimizer tags
+    sm_list = [':'.join([m['software'], m['minimizer']])
+               for m in [dict(zip(col_order, t.split(":")))
+               for t in list(column_summary.keys())]]
+
+    # Find col tags with each sm_tag
+    sm_summary = {}
+    for sm in sm_list:
+        tags = set()
+        for tag in all_result_tags:
+            up = dict(zip(col_order, tag["col"].split(":")))
+            if ':'.join([up['software'], up['minimizer']]) == sm:
+                tags.add(tag['col'])
+        sm_summary[sm] = tags
```

Comment:
Is sm_list only used to generate the sm_summary? If yes, we can eliminate a loop by not generating this first.

I'd also suggest using tuples (will need to avoid the other places where it is joined and split)

Comparison works just the same with tuples
```
('foo', 'bar') == ('foo', 'bar') #True
('hello', 'world') == ('foo', 'bar') #False
```


```suggestion
    # Find col tags with each sm_tag
    sm_summary = {}
    for tag in all_result_tags:
        up = tag["col"].split(":")
        sm = (up[col_order.index('software')],
              up[col_order.index('minimizer')])
        if sm not in sm_summary:
           sm_summary[sm] = set()
        sm_summary[sm].add(tag['col'])
```